### PR TITLE
Update JS API spec to V2.

### PR DESCRIPTION
**Pull request by zolkis**
Fixes a couple of issues with the JS API vs the 1.0 version of the OIC Core Specification, such as:
- addressing of an OicResource becomes deviceId + url (path)
- remove args from findDevices() as it does an endpoint discovery
- remove updatedPropertyNames (no support in OIC)
- use OicResource as parameters to functions, though only a few properties are used by the implementation (mainly deviceId and url)
- remove bogus event handlers from OicClient
- remove connectionMode (implementation detail)
- add list of OicResources to OicDeviceFoundEvent (though only url, resourceType and interfaces are provided)
- updated example.

The new API spec does all the above, plus:
- reorganize methods to better reflect capabilities of OicDevice vs OicResource, and make a more natural development flow with these objects
- add possibility to install client side handlers to delete and update events
- add possibility to install separate handlers to different server side CRUDN requests
- update and improve examples reflecting the new capabilities
- update the full text of the API spec
- add algorithmic descriptions to findDevices() and findResources().

### CODE REVIEW

File: spec/iot-js-spec.md [hunk 1]
```diff
@@ -160,7 +153,6 @@ interface OicRequestEvent : Event {
   readonly attribute USVString source;  // device uuid of the OIC client making the request
   readonly attribute USVString target;  // id of the target resource
   readonly attribute OicResourceRepresentation? res;  // for “create” and “update” only
```
Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: You need to change the comment to say ```// for "update" only``` and you also need to add
```
readonly attribute OicResourceInit? init; // for "create" only
```
because in the case of create we need the metadata as well (url, the types, the interfaces, and the discoverable/observable flags).

Comment (zolkis) [inline spec/iot-js-spec.md]: Actually it should be an OicResourceInit in both cases. Thanks.

File: spec/iot-js-spec.md [hunk 2]
```diff
@@ -183,23 +171,21 @@ interface OicResourceFoundEvent : Event {
 
 interface OicDeviceFoundEvent : Event {
   OicDeviceInfo device;
+  sequence<OicResource> resources;  // url, resourceType, interfaces
 };
 
 dictionary OicDiscoveryOptions {  // all properties are null by default, meaning “find all”
   USVString? deviceId;
-  USVString? resourceId;
+  USVString? resourceURL;
```
Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: Please make this ResourceUrl ... the fewer caps, the better.

Comment (zolkis) [inline spec/iot-js-spec.md]: We had a round on this with Domenic, and we have to use URL...
See https://w3ctag.github.io/design-principles/

File: spec/iot-js-spec.md [hunk 3]
```diff
@@ -98,11 +95,9 @@ dictionary OicDeviceSettings {
   USVString url;  // host:port
   OicDeviceInfo info;
   OicDeviceRole role;
-  OicConnectionMode connectionMode;  // mapping to “Quality of Service” in IoTivity
```
Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: Why did you remove this? This is easy enough to implement.

File: spec/iot-js-spec.md [hunk 4]
```diff
@@ -123,31 +118,29 @@ dictionary OicDeviceInfo {
...
 
   // client API: CRUDN
-  Promise<OicResource> createResource(OicResourceInit resource);  // create remote res
-  Promise<OicResource> retrieveResource(USVString resourceId);  // get remote res
-  Promise<void> updateResource(USVString resourceId, OicResourceInit resource);  // post or put remote res
-  Promise<void> deleteResource(USVString resourceId);  // delete remote res
-  Promise<OicResource> startObserving(USVString resourceId);  // retrieve + observe flag
-  Promise<void> cancelObserving(USVString id);
-
-  attribute EventHandler<OicResourceChangeEvent> onresourcechange;  // observe
```
Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: Umm, if you remove onresourcechange, how will you do observation?

File: spec/iot-js-spec.md (hunk 3 above)
Comment (zolkis) [inline spec/iot-js-spec.md]: It is implementation detail until we have a use case for it in clients. We can add it back any time. If you have a use case, could add it back now :).

File: spec/iot-js-spec.md (hunk 4 above)
Comment (zolkis) [inline spec/iot-js-spec.md]: You can choose having one of the following:
- onresourceupdate
- startObserving() and cancelObserving().

Right now it has the latter, but I'd like the former.

Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: node.js has an [EventEmitter](https://nodejs.org/api/events.html#events_event_newlistener) API which lets me know whenever a new listener is attached to an object, and how many listeners are currently attach. That lets me do special things like initiating observation when someone attaches the first handler and cancelling observation when someone detaches the last handler. Wonderful! However, you gotta ask yourself if Tizen and Crosswalk also have this capability.

Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: So, I guess what I'm trying to say is that, before we go for one xor the other, let's make sure we don't need both on any of the platforms we seek to target.

Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: Also, where exactly is onresourceupdate? I'm not seeing it on https://github.com/zolkis/iotivity-node/blob/master/spec/iot-js-spec.md ... or did you mean you were gonna add one xor the other?

Comment (zolkis) [inline spec/iot-js-spec.md]: I would like to rename from onresourcechange to onresourceupdate, since that is more in line with the Core spec.
And no, we don't need startObserve() and cancelObserve(), since we can hook to the EventHandler anyway. It was my first uneducated attempt to support observe, let's not make a legacy out of it :).

**Comment by hekra01** [comment]
>remove args from findDevices() as it does an endpoint discovery

Then we dont have the possiblity to do unicast device discovery as in OCPlatform::getDeviceInfo(const std::string & host, ...)

File: spec/iot-js-spec.md (hunk 4 above)
Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: OK, if all our target platforms support custom actions during event handler attachment/detachment, then I'm OK with an event-only approach. Nevertheless, now that I think about it, I believe that if we do onresourceupdate, it's only possible in node.js if we add the event to OicResource. The reason is that yes, I can do something special when the first event handler is attached, and yes, I can do something special when the last event handler is detached, but I can only do something special concerning the object to which it is attached/detached. If that object is device.client, I won't know which resource to start observing or which resource to stop observing. So, the event would have to be attached to an OicResource for this to work.

Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: So, this is going to be a step towards the new spec.

Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: ... which means that we should leave onresourcechange, startObserving(), and cancelObserving() intact for the short term (unfortunately), and implement it the Right Way™ in the long term, based on the new spec.

**Comment by zolkis** [comment]
I have looked in the C++ API implementation, and there are other unfulfilled promises there, including for instance unicast device discovery: the API makes it look like it is possible, but as far as I could find, it is not implemented. Nor could I find it supported in the Core spec. If it is HTTP discovery ([mDNS](https://tools.ietf.org/html/rfc6762)), then it is multicast. If it is [CoAP discovery](https://tools.ietf.org/html/rfc7252#section-7), the OIC Core spec says it is multicast. So I don't see support there for unicast device discovery.

Should I include it back "for any case" as an optional arg, or should I cut it out until we bump into the lack of it and add only then?
Usually it's easier to add than remove things.

File: spec/iot-js-spec.md (hunk 2 above)
Comment (zolkis) [inline spec/iot-js-spec.md]: I changed it to resourcePath. Looks like it *is* used like a path only, always.

File: spec/iot-js-spec.md (hunk 4 above)
Comment (zolkis) [inline spec/iot-js-spec.md]: We have implemented hooks on addEventListener in Crosswalk extensions. AFAIK this can be done everywhere. The exception is when you *attribute* a function to ```onresourceupdate```. Then you have to make the hook from native. But it is normal course of events.

Comment (zolkis) [inline spec/iot-js-spec.md]: Anyway, that indeed belongs to the new API. With this one we need to keep the functions as well.

Comment (gabrielschulhof) [inline spec/iot-js-spec.md]: Actually, you don't need native for hooking into onresourceupdate because Javascript has [property accessors](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty) which are supported everywhere except IE8.

**Comment by hekra01** [comment]
> I have looked in the C++ API implementation, and there are some unfulfilled promises there, including unicast device discovery: the API makes it look like it is possible, but as far as I could find, it is not implemented

It tested it successfully. It works if port is provided in host. 
A jira issue already exists about you observation, I commented there also: https://jira.iotivity.org/browse/IOT-823

> Nor could I find it supported in the Core spec

Core spec v0.9.9 at 11.3.2.2 Direct discovery:
"The request may be made as a unicast or 2017 multicast."

There is also a test case for this in OIC_Compliance_and_Interoperability_Test_Specification_v0 58: 
"CT1.1.5 CoAP Based Unicast Discovery on /oic/p"

**Comment by zolkis** [comment]
It seems that the gap between the old (v1) and new (v2) API is closing. v2 moves more client methods to OicResource, making programming more clear and more intuitive.

A question about resource id. Right now we use {deviceId, resourcePath} tuple. Usually in Web API's they use URLs in this case. At least @kenchris is in favor of that. I wonder if should we use URLs (exposed as USVString) with the following notation:
http://```<address:port>```/```<path>``` means the device id is specified as address:port tuple,
and
oic://```<deviceIdString>```/```<path>``` means the device id if an opaque string, and OIC is used for identifying the device.
I emphasize this is a *client abstraction* for the JS API, not the OIC Core spec URL semantics. Implementations and clients would need to parse that string in order to separate resource path from device id.
Leaving the {deviceId, path } tuple also works, albeit not so web'ish.
I am leaning towards leaving this as a tuple. Opinions?

**Comment by gabrielschulhof** [comment]
I vote tuple.

**Comment by poussa** [comment]
I vote for tuple as well. oic scheme looks like overloading it.

**Comment by hekra01** [comment]
> I think we should add "secure" to the ConnectionMode

At least 2 cases for security:
* First, adding `OicResourceInit.secure` to use for secure resource registration as in
```
OCPlatform::registerResource(handle, uri, type, interface, cb, OC_DISCOVERABLE | OC_OBSERVABLE | OC_SECURE);
```
* Second, secure connection: discussed before.

> PlatformConfig in IoTivity in fact should be device config IMHO.

OK, but then, I observe that when the JS spec/implementations are moved to the IoTivity repos, their semantics of platform/device management will differ from the ones used by the C/C++/Java APIs.

**Comment by zolkis** [comment]
I will rework the connectivity and security related parts. Thanks for the hints.
However, it is not a problem if the JS API slightly differs from the native ones.

**Comment by hekra01** [comment]
One other use case with security is the provisioning part: ownership, ACLs and certificates, etc... need to be managed, for secured IoTivity devices to communicate. (all details in OIC security spec)

`OCProvisioningManager.h` provides the APIs for this, and provisioning can be done from command line (provisioningclient.c or provisioningclient.cpp) 

Maybe this part would make sense for addition to the JS APIs? (Use case: configure device provisioning from web application)

**Comment by poussa** [comment]
Provisioning and security must be part of the API.

**Comment by zolkis** [comment]
I suggest we include provisioning in the next version, v3.
We should still leave this PR open until v2, and changes to v1 are finalized.

**Comment by zolkis** [comment]
With the latest developer feedback and latest changes, the differences between V1 and V2 APIs became minor. Now the proposed version is called the V2 API. Updated the PR name.

**Comment by zqzhang** [comment]
> We should still leave this PR open until v2, and changes to v1 are finalized.

Do we finish this? Or is there anything block this PR merging?

**Comment by zolkis** [comment]
We should still leave this open. Implementation of the spec is ongoing, and we need that feedback to this PR. Also, we play with this in the REST API server and get feedback from there, too. We leave the spec open until it is fairly stable. Looks like soon (this week?) we could merge.

File: spec/iot-js-spec.md [hunk 5]
```diff
@@ -4,331 +4,476 @@ IoT Web API
...
-  OicResource resource;
-  DOMString[] updatedPropertyNames; // only for updates, for the rest empty
+### OIC Client
+The OIC Client API provides functionality to access (CRUD) remote resources.
+```javascript
+[NoInterfaceObject]
+interface OicClient {
+  Promise<OicResource> create(OicResourceInit resource);
+  Promise<void> retrieve(OicResourceId id, optional Dictionary parameters);
+  Promise<void> update(OicResourceInit resource);  // partial dictionary
```
Comment (nagineni) [inline spec/iot-js-spec.md]: What does it mean by 'partial dictionary'? We should pass the 'resource' with updated properties.

File: spec/iot-js-spec.md [hunk 6]
```diff
@@ -4,331 +4,476 @@ IoT Web API
...
-  enum { “add”, “update”, “remove” } type;
-  OicResource resource;
-  DOMString[] updatedPropertyNames; // only for updates, for the rest empty
+### OIC Client
+The OIC Client API provides functionality to access (CRUD) remote resources.
+```javascript
+[NoInterfaceObject]
+interface OicClient {
+  Promise<OicResource> create(OicResourceInit resource);
+  Promise<void> retrieve(OicResourceId id, optional Dictionary parameters);
```
Comment (nagineni) [inline spec/iot-js-spec.md]: Thanks for adding parameters as optional. We need it to pass the query parameters to retrieve.

File: spec/iot-js-spec.md [hunk 7]
```diff
@@ -4,331 +4,476 @@ IoT Web API
...
-  USVString? resourceId;
-  DOMString? resourceType;
+  // handle CRUDN requests from clients
+  attribute EventHandler<OicRequestEvent> onobserverequest;
+  attribute EventHandler<OicRequestEvent> onunobserverequest;
+  attribute EventHandler<OicRequestEvent> onretrieverequest;
+  attribute EventHandler<OicRequestEvent> ondeleterequest;
+  attribute EventHandler<OicResourceEvent>  onupdaterequest;
+  attribute EventHandler<OicResourceEvent>  oncreaterequest;
+
```
Comment (nagineni) [inline spec/iot-js-spec.md]: :+1:

File: spec/iot-js-spec.md (hunk 5 above)
Comment (zolkis) [inline spec/iot-js-spec.md]: We need to check that - it would not be very efficient to send the whole resource all again even if one property was changed. The OIC Core spec somehow is not very clear on that.

File: spec/iot-js-spec.md [hunk 8]
```diff
@@ -4,331 +4,476 @@ IoT Web API
...
 
-  - an installable software module (called *application*) is deployed in a gateway or dedicated solution box on the field, and
+**Resource identification** is URL path, relative to a given device.
 
-  - a cloud service with user accounts and profiles, remote management etc.
+On each device there are special resources, implementing device discovery, resource discovery, platform discovery, etc. Platform needs to be discoverable on a resource with a fixed URI ```/oic/p```, device on ```/oic/d``` and resources on ```/oic/res```. This API encapsulates these special resources and the hardcoded/fixed URIs by explicit function names and parameters.
 
-Multiple providers/solutions can share a gateway (a.k.a. OIC intermediary device);
-gateways' support for sensors can be extended via a software module as part of the solution.
+**Device discovery** uses endpoint discovery: multicast request "GET /oic/res" to "All CoAP nodes" (```224.0.1.187``` for IPv4 and ```FF0X::FD``` for IPv6, port 5683). The response lists devices and their resources (at least URI, resource type, interfaces, and media types). Since this is basically a resource discovery, it is merged with resource discovery.
```
Comment (hekra01) [inline spec/iot-js-spec.md]: > **Device discovery** uses endpoint discovery: multicast request "GET /oic/res"

hm, we will use `oic/d` and not `oic/res?rt=oic.wk.d` for device discovery, right?

Comment (zolkis) [inline spec/iot-js-spec.md]: This tries to be an informative section and it was distilled from the Core spec. What happens exactly depends on the underlying implementation; my understanding is the oic/d resource is mandatory to implement, and it is discovered by the query you mentioned. BTW the last sentence is not correct in this version of the API - I need to fix this.

**Comment by poussa** [comment]
@zolkis Should we merge this since it has all been implemented in the code already, right?

**Comment by zolkis** [comment]
Yes, we could merge this.

**Comment by zqzhang** [comment]
@gabrielschulhof could you help merge this per the above comments?
